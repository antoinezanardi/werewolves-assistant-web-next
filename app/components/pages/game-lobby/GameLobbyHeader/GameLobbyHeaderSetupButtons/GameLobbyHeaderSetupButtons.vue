<template>
  <PrimeVueButtonGroup>
    <TransitionGroup
      name="fade-list"
    >
      <GameLobbyHeaderOptionButton
        id="game-lobby-header-options-button"
        key="game-options-button"
        class="fade-list-item"
        @game-options-button-click="onGameOptionsButtonClickFromGameOptionButton"
      />

      <GameLobbyHeaderPositionCoordinatorButton
        v-if="isPositionCoordinatorVisible"
        id="game-lobby-header-position-coordinator-button"
        key="game-position-coordinator-button"
        ref="gameLobbyHeaderPositionCoordinatorButton"
        class="fade-list-item"
        @position-coordinator-button-click="onPositionCoordinatorButtonClickFromGamePositionCoordinatorButton"
      />

      <GameLobbyHeaderAdditionalCardsManagerButton
        v-if="isAdditionalCardsManagerVisible"
        id="game-lobby-header-additional-cards-manager-button"
        key="game-additional-cards-manager-button"
        class="fade-list-item"
        @additional-cards-manager-button-click="onAdditionalCardsManagerButtonClickFromGameAdditionalCardsManagerButton"
      />
    </TransitionGroup>
  </PrimeVueButtonGroup>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";
import type { ComponentPublicInstance } from "vue";
import type { GameLobbyHeaderSetupButtonsEmits, GameLobbyHeaderSetupButtonsExposed } from "~/components/pages/game-lobby/GameLobbyHeader/GameLobbyHeaderSetupButtons/game-lobby-header-setup-buttons.types";
import GameLobbyHeaderAdditionalCardsManagerButton from "~/components/pages/game-lobby/GameLobbyHeader/GameLobbyHeaderSetupButtons/GameLobbyHeaderAdditionalCardsManagerButton/GameLobbyHeaderAdditionalCardsManagerButton.vue";
import GameLobbyHeaderOptionButton from "~/components/pages/game-lobby/GameLobbyHeader/GameLobbyHeaderSetupButtons/GameLobbyHeaderOptionsButton/GameLobbyHeaderOptionsButton.vue";
import GameLobbyHeaderPositionCoordinatorButton from "~/components/pages/game-lobby/GameLobbyHeader/GameLobbyHeaderSetupButtons/GameLobbyHeaderPositionCoordinatorButton/GameLobbyHeaderPositionCoordinatorButton.vue";
import { useAnimateCss } from "~/composables/animate-css/useAnimateCss";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";

const emit = defineEmits<GameLobbyHeaderSetupButtonsEmits>();

const createGameDtoStore = useCreateGameDtoStore();
const { createGameDto, doesCreateGameDtoContainAdditionalCardsDependantRoles } = storeToRefs(createGameDtoStore);

const minPlayerToDisplayPositionCoordinator = 2;

const { animateElementOnce } = useAnimateCss();

const gameLobbyHeaderPositionCoordinatorButton = ref<ComponentPublicInstance | null>(null);

const isPositionCoordinatorVisible = computed<boolean>(() => createGameDto.value.players.length >= minPlayerToDisplayPositionCoordinator);

const isAdditionalCardsManagerVisible = computed<boolean>(() => doesCreateGameDtoContainAdditionalCardsDependantRoles.value);

function onGameOptionsButtonClickFromGameOptionButton(): void {
  emit("gameOptionsButtonClick");
}

function onPositionCoordinatorButtonClickFromGamePositionCoordinatorButton(): void {
  emit("positionCoordinatorButtonClick");
}

function onAdditionalCardsManagerButtonClickFromGameAdditionalCardsManagerButton(): void {
  emit("additionalCardsManagerButtonClick");
}

async function highlightPositionCoordinatorButton(): Promise<void> {
  if (!gameLobbyHeaderPositionCoordinatorButton.value) {
    throw createError("Game Lobby Header Position Coordinator Button is not defined");
  }
  await animateElementOnce((gameLobbyHeaderPositionCoordinatorButton.value.$el as HTMLElement), "heartBeat");
}

defineExpose<GameLobbyHeaderSetupButtonsExposed>({
  highlightPositionCoordinatorButton,
});
</script>