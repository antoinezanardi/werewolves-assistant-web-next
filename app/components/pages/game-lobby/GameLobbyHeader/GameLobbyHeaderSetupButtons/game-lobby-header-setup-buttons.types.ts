type GameLobbyHeaderSetupButtonsEmits = {
  gameOptionsButtonClick: [];
  positionCoordinatorButtonClick: [];
  additionalCardsManagerButtonClick: [];
  groupOrganizerButtonClick: [];
};

type GameLobbyHeaderSetupButtonsExposed = {
  highlightGameOptionsButton: () => void;
  highlightPositionCoordinatorButton: () => void;
  highlightAdditionalCardsManagerButton: () => void;
};

export type {
  GameLobbyHeaderSetupButtonsEmits,
  GameLobbyHeaderSetupButtonsExposed,
};