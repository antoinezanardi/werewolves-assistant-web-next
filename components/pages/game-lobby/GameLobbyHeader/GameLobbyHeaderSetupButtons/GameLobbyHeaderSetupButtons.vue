<template>
  <VuePrimeButtonGroup>
    <GameLobbyHeaderOptionButton
      id="game-lobby-header-options-button"
      @game-options-button-click="handleGameOptionsButtonClick"
    />
  </VuePrimeButtonGroup>
</template>

<script setup lang="ts">
import type { GameLobbyHeaderSetupButtonsEmits } from "~/components/pages/game-lobby/GameLobbyHeader/GameLobbyHeaderSetupButtons/game-lobby-header-setup-buttons.types";
import GameLobbyHeaderOptionButton from "~/components/pages/game-lobby/GameLobbyHeader/GameLobbyHeaderSetupButtons/GameLobbyHeaderOptionsButton/GameLobbyHeaderOptionsButton.vue";

const emit = defineEmits<GameLobbyHeaderSetupButtonsEmits>();

function handleGameOptionsButtonClick(): void {
  emit("gameOptionsButtonClick");
}
</script>