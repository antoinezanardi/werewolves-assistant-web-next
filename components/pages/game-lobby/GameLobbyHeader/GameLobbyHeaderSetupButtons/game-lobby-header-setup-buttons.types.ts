type GameLobbyHeaderSetupButtonsEmits = {
  gameOptionsButtonClick: [];
};

export type { GameLobbyHeaderSetupButtonsEmits };