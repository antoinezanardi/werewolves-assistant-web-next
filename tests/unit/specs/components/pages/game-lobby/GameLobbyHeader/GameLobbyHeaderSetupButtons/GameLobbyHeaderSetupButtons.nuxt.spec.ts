import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";
import type GameLobbyHeaderOptionsButton from "~/components/pages/game-lobby/GameLobbyHeader/GameLobbyHeaderSetupButtons/GameLobbyHeaderOptionsButton/GameLobbyHeaderOptionsButton.vue";
import GameLobbyHeaderSetupButtons from "~/components/pages/game-lobby/GameLobbyHeader/GameLobbyHeaderSetupButtons/GameLobbyHeaderSetupButtons.vue";

import { mountSuspendedComponent } from "~/tests/unit/utils/helpers/mount.helpers";
import type { VueVm } from "~/tests/unit/utils/types/vue-test-utils.types";

describe("Game Lobby Header Setup Buttons Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameLobbyHeaderSetupButtons>>;

  async function mountGameLobbyHeaderSetupButtonsComponent(options: ComponentMountingOptions<typeof GameLobbyHeaderSetupButtons> = {}):
  Promise<ReturnType<typeof mount<typeof GameLobbyHeaderSetupButtons>>> {
    return mountSuspendedComponent(GameLobbyHeaderSetupButtons, {
      global: { stubs: { ButtonGroup: false } },
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameLobbyHeaderSetupButtonsComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Options Button", () => {
    it("should emit 'gameOptionsButtonClick' event when clicked.", () => {
      const optionsButton = wrapper.findComponent<typeof GameLobbyHeaderOptionsButton>("#game-lobby-header-options-button");
      (optionsButton.vm as VueVm).$emit("gameOptionsButtonClick");

      expect(wrapper.emitted("gameOptionsButtonClick")).toBeTruthy();
    });
  });
});