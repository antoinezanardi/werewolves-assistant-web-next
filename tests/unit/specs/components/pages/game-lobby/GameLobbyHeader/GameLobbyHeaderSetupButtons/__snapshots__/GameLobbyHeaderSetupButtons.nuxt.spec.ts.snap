// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Lobby Header Setup Buttons Component > should match snapshot when rendered. 1`] = `"<span class="p-button-group p-component" role="group" data-pc-name="buttongroup" data-pc-section="root"><game-lobby-header-option-button-stub id="game-lobby-header-options-button"></game-lobby-header-option-button-stub></span>"`;
